### PR TITLE
Unit test queryAll

### PR DESCRIPTION
**Pull request by inancgumus**
This one depends on #225, and it's a part of the tasks in #215.

Things to remember before undrafting.

- [x] Remove: 70063cb

Abstract away the following first:

- [x] h.evalWithScript

Tests:

- [x] Error with an invalid selector
- [x] Error when it cannot evaluate the javascript
- [x] Nil result without an error
- [x] Catches the invalid JSHandle type

Further Tests:

- [x] Disposes the returned element
- [x] Disposes the handles (_should only dispose non-nil ones_)
- [x] Receives correct function: js.QueryAll
- [x] Receives correct evalOptions

### CODE REVIEW

**Comment by inancgumus** [comment]
@imiric, Could you take a look at how it looks so far? Thanks.

**Review by imiric — COMMENTED** [review]
LGTM, just some minor suggestions.

Though I do wonder if we should bother with unit testing `ElementHandle` methods at this level. Their behavior is very much dependent on a JS environment and the DOM, and it takes quite a lot of changes to make stubbing possible, all to test ~30 LOCs.

The current [`TestElementHandleQueryAll`](https://github.com/grafana/xk6-browser/blob/390bb444d84fe93edc9eac252b4e74c9bdb87fa9/tests/element_handle_test.go#L10) seems like a much more valuable test, that is easy to read and write, and requires no complex test doubles.

I think we should put this effort on pause and focus on higher-level functional tests, at least for `ElementHandle`.

**Comment by inancgumus** [comment]
Thanks for your review.

I agree with you to an extent. High-level tests are more readable, however: 
* How can we detect and prevent regressions (if any) if code doesn't dispose of acquired handles?
* What happens, and how does it behave if it cannot evaluate?
* Whether it returns nil handles or not?
* Is it easier to find errors in high-level tests compared to unit tests?
* ...

We can do some of them to an extent with high-level code, and will they be clearer? Can we use the same stub within other tests?

File: common/element_handle.go [hunk 1]
```diff
@@ -62,6 +62,10 @@ var _ api.JSHandle = &ElementHandle{}
 type (
 	elementHandleActionFn        func(context.Context, *ElementHandle) (interface{}, error)
 	elementHandlePointerActionFn func(context.Context, *ElementHandle, *Position) (interface{}, error)
+
+	// evalFunc is currently one of evalWithScript and eval.
```
[the remark this answers is not in the document]
Comment (inancgumus) [reply common/element_handle.go:66]:
```suggestion
// evalFunc is a common interface for both evalWithScript and eval.
```

File: common/element_handle.go [hunk 2]
```diff
@@ -1206,26 +1210,38 @@ func (h *ElementHandle) Query(selector string) api.ElementHandle {
 // QueryAll queries element subtree for matching elements.
 // If no element matches the selector, the return value resolves to "null".
 func (h *ElementHandle) QueryAll(selector string) []api.ElementHandle {
+	defer applySlowMo(h.ctx)
```
[the remark this answers is not in the document]
Comment (inancgumus) [reply common/element_handle.go:1213]: I think you're right, see #234.

**Comment by imiric** [comment]
While those points are important, I don't think they're as important _right now_ as proving that basic functionality works. What good is to users that we prove that the usage of `Dispose()` works correctly, if essential things like [HTTP basic auth](https://github.com/grafana/xk6-browser/issues/203) don't work and are not tested? We can worry about code coverage after we have these higher-level tests, at which point we can drill down and write tests for these smaller units.

Essentially we should tackle the [test pyramid](https://martinfowler.com/articles/practical-test-pyramid.html) with a top-down approach. Right now we have a few E2E tests, and mostly integration (browser) tests. Instead of focusing on unit tests, we should work on fixing basic functionality, adding more integration and some E2E tests, and _when possible_ unit tests. But unit tests shouldn't be our top priority right now.

**Review by imiric — APPROVED** [review]

**Comment by inancgumus** [comment]
I think all of these test types are important to have (unit+functional+e2e), and we should tackle them together, step by step, instead of separately prioritizing them in different stages.

We can see whether there is a problem in the system or it works correctly using a higher-level test (like functional and e2e). However, we won't see the cause of the failure. And it will take us longer to figure it out. We experienced this first hand. It's pretty hard to solve bugs when you don't have unit tests.

Although testing the user-facing functionality is critical, having a quality codebase for its developers is also valuable. Not having unit tests will accrue technical debt in the codebase. Adding features by accruing technical debt is:
* Dangerous.
* It will slow us down each passing day.
* It will also alienate newcomers to our team and make their onboarding process painful and longer.
* It will be harder to add unit tests later on because higher-level tests don't mandate anything about code structure, and there are a lot of dependencies in every component. This is why it is harder to unit test right now, and it will be worse if we don't do it.

**Comment by inancgumus** [comment]
> Essentially we should tackle the [test pyramid](https://martinfowler.com/articles/practical-test-pyramid.html) with a top-down approach.

It's called _"ice cream cone testing"_ and I don't think it's a good way of testing software.

Here's my take. We should keep working on:
* Fixing the basic functionality
* Adding unit tests and refactoring the codebase every day
* Adding more integration tests
* Some E2E tests

All of these test types are important to have (unit+functional+e2e), and we should tackle them together, step by step, instead of separately prioritizing them in different stages.

We can see whether there is a problem in the system or if it works correctly using a higher-level test (like functional and e2e). However, we won't easily see the cause of the failure, so it will take us longer to figure it out. We experienced this first hand. It's pretty hard to solve bugs when you don't have unit tests.

Although testing the user-facing functionality is critical, having a quality codebase for its developers is also valuable. Not having unit tests will accrue technical debt in the codebase. Adding features by accruing technical debt is dangerous because:

* It will slow us down each passing day.
* It will also alienate newcomers to our team and make their onboarding process painful and longer.
* It will be harder to add unit tests later on because higher-level tests don't mandate anything about code structure, and there are a lot of dependencies in every component (and will be more). This is why it is harder to unit test right now, and it will be worse if we don't do it.

---

Read more: https://martinfowler.com/bliki/TestPyramid.html

From the Google testing blog ([read here](https://testing.googleblog.com/2015/04/just-say-no-to-more-end-to-end-tests.html)):

> Try to avoid these anti-patterns:
> 1. Inverted pyramid/ice cream cone. The team relies primarily on end-to-end tests, using few integration tests and even fewer unit tests. 
> 2. Hourglass. The team starts with a lot of unit tests, then uses end-to-end tests where integration tests should be used. The hourglass has many unit tests at the bottom and many end-to-end tests at the top, but few integration tests in the middle.

**Comment by imiric** [comment]
> It's called "ice cream cone testing" and I don't think it's a good way of testing software.

It's not what I'm suggesting, but let's discuss it next week.

**Pull request merged**